### PR TITLE
Azure Firewall - Data PIP removal for FT firewal

### PR DESCRIPTION
**Pull request by tinawu6**
## Description

Feature: Enable firewall to be created and reconfigured without customer providing public IP.
This change relaxes the validation that enforces user to add a data PIP when creating a firewall. 
The new supported scenarios are: 
- Allowing user to create FT firewall without data PIP
- Allowing user to remove data PIP from FT firewall

## Checklist

- [X] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [X] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage

### CODE REVIEW

File: src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 1]
```diff
@@ -99,10 +99,10 @@ public class NewAzureFirewallCommand : AzureFirewallBaseCmdlet
         public PSVirtualNetwork VirtualNetwork { get; set; }
 
         [Parameter(
-            Mandatory = true,
+            Mandatory = false,
             ValueFromPipelineByPropertyName = true,
             ParameterSetName = "IpConfigurationParameterValues",
-            HelpMessage = "One or more Public IP Addresses. The Public IP addresses must use Standard SKU and must belong to the same resource group as the Firewall.")]
+            HelpMessage = "Allowing zero data Public IP for FT. The Public IP addresses must use Standard SKU and must belong to the same resource group as the Firewall.")]
```
Comment (iatodoro) [inline src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs:105]: The original help message is still accurate - the value of this parameter should be One or more PIPs

File: src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs [hunk 2]
```diff
@@ -174,14 +177,21 @@ public void Allocate(PSVirtualNetwork virtualNetwork, PSPublicIpAddress[] public
 
             this.IpConfigurations = new List<PSAzureFirewallIpConfiguration>();
 
-            for (var i = 0; i < publicIpAddresses.Count(); i++)
+            if (publicIpAddresses != null && publicIpAddresses.Count() != 0)
```
Comment (iatodoro) [inline src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs:180]: Should be `> 0` to be entirely semantically correct.

File: src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs [hunk 3]
```diff
@@ -252,21 +271,37 @@ public void RemovePublicIpAddress(PSPublicIpAddress publicIpAddress)
                 throw new ArgumentException($"Public IP Address {publicIpAddress.Id} is not attached to firewall {this.Name}");
             }
 
-            if (this.IpConfigurations.Count > 1 && ipConfigToRemove.Subnet != null)
-            {
-                throw new InvalidOperationException($"Cannot remove IpConfiguration {ipConfigToRemove.Name} because it references subnet {ipConfigToRemove.Subnet.Id}. Move the subnet reference to another IpConfiguration and try again.");
+            if (this.ManagementIpConfiguration != null) {
+                if (ipConfigToRemove.Subnet != null)
+                {
+                    Console.ForegroundColor = ConsoleColor.Yellow;
```
Comment (iatodoro) [inline src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs:277]: This message is not necessary - the Public IP is not associated with the subnet in any way.

File: src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs [hunk 4]
```diff
@@ -252,21 +271,37 @@ public void RemovePublicIpAddress(PSPublicIpAddress publicIpAddress)
...
+                    Console.WriteLine($"Removing the Public IP Address associated with subnet {ipConfigToRemove.Subnet.Id}.");
```
Comment (iatodoro) [inline src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs:278]: This message is not necessary - the Public IP is not associated with the subnet in any way.

File: src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs [hunk 5]
```diff
@@ -252,21 +271,37 @@ public void RemovePublicIpAddress(PSPublicIpAddress publicIpAddress)
...
 
-                this.ManagementIpConfiguration = null;
-            }
+                if (this.IpConfigurations.Count == 1)
+                {
+                    Console.ForegroundColor = ConsoleColor.Yellow;
+                    Console.WriteLine($"WARNING: Removing the last Public IP Address, this will deallocate the firewall. You will have to invoke {nameof(Allocate)} to reallocate it.");
+                    Console.ResetColor();
+
+                    this.ManagementIpConfiguration = null;
```
Comment (iatodoro) [inline src/Network/Network/Models/AzureFirewall/PSAzureFirewall.cs:300]: No longer needed as you're in the `else` part where ManagementIpConfiguration will always be null.

**Comment by isra-fel** [comment]
As we talked in mail, please use the link https://aka.ms/AzureGithub to join Azure orgnization on Github. It marks you as a trusted developer so the CI pipelines can trigger automatically. @tinawu6

**Review by isra-fel — CHANGES_REQUESTED** [review]
Hi @tinawu6 , changing cmdlet syntax requires you to [regenerate help document](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)
In this case, you should expect [New-AzFirewall.md](https://github.com/Azure/azure-powershell/blob/master/src/Network/Network/help/New-AzFirewall.md) to be updated after regeneration.
Other changes look good to me

**Review by isra-fel — APPROVED** [review]

**Review by VeryEarly — APPROVED** [review]

**Pull request merged**